### DISCUSSION TITLE
Does AroCC support outputting C code ?

### DISCUSSION THREAD

**Discussion by heysokam** [Vexu/arocc#715] [Q&A] [answered]
I'm writing a transpiler for another lang in Zig, and I was wondering if AroCC has the ability to convert its Tokens or AST into valid C code at some point in its process.
The idea would be to use AroCC as a library, generate the data expected by AroCC manually in my transpiler, and then send it to AroCC in a form it understands.

Is C source code output possible with the existing toolset of AroCC in some way?

**Comment by Vexu** [comment]
It does not since it doesn't implement a C formatter. I'm also not sure something like that is even needed for outputting C. [The Zig compiler's C backend](https://github.com/ziglang/zig/blob/e1d4cf67caedfe20d2e0af9d3b518af1f0777191/src/codegen/c.zig) just appends text to some buffers. You could probably find useful parts to copy from there to get started.

**Reply by heysokam** [reply]
Somebody else sent me to the C backend. And I landed there myself before they sent me to the file.
But its like my first week in Zig, and my first real compiler project ever... so getting deep into a 7k sloc file, right at the deepest depths of the Zig compiler, just to know if the fully mangled C output of the Zig.C backend is even usable at all in some partial way is not very approachable at all.

The main issue I had reading that file is that I have no clue how to tell if the mangling is happening as an implicit/mandatory process of that file _(which I tried to understand a couple of times before you sent me there, I was aware of it)_ or if the Zig compiler is doing that mangling somewhere else in its really difficult to follow AST and compilation process.

Do you know anything about that mangling?
How could I tell if the C backend is creating its Zig-like code there in that file, or if that's just a code generator/formatter and the Zig-to-C is happening somewhere else?

**Reply by Vexu** [reply]
If you're talking about identifier mangling then that is required because C has reserved identifiers (keywords, type names, library functions) and because Zig allows arbitrary byte sequences to be used as identifiers with the `@""` syntax.

And if the size is daunting you could always start by looking at an older, simpler version of the C backend like [the version shipped with 0.9.0](https://github.com/ziglang/zig/blob/a18bf7a7bfa4e8c32aa25295cfa1ca768e5f5b74/src/codegen/c.zig) (note though that it is missing some major optimizations like https://github.com/ziglang/zig/pull/13768.

**Reply by heysokam** [reply]
Good point, didn't think of looking at earlier versions. Might even be good to take a look at early commits in whatever PR introduced the feature, just to see how the code evolved in the early days.
Good thinking :writing_hand:

**Reply by heysokam** [reply, marked as answer]
The Zig's C backend doesn't seem to be usable for non-mangled output.
From my limited understanding so far, I believe their formatter functions are what mangle the resulting symbols.
